### ISSUE TITLE
[pyQGIS] QgsRasterBlock fails if not every cell is set

### ISSUE THREAD

**Issue by moovida** [qgis/QGIS#53062]
### What is the bug or the crash?

When I create a QgsRasterBlock and use setValue to set a value in a given position, without setting every cell (which kind of defeates the purpose of setValue), the resulting raster is scrambled.

### Steps to reproduce the issue

Load any single band elevation model and create from that a new constant raster:

```
newRaster = processing.run("native:createconstantrasterlayer", {
    'EXTENT': dtmLayer.extent(),
    'TARGET_CRS': crs,
    'PIXEL_SIZE':dtmLayer.rasterUnitsPerPixelX(),
    'NUMBER':-9999.0,
    'OUTPUT_TYPE': 6,
    'OUTPUT':'TEMPORARY_OUTPUT'
    }
)


newRasterLayer = QgsRasterLayer(newRaster['OUTPUT'], 'temp', 'gdal')
newRasterProvider = newRasterLayer.dataProvider()
```
then loop over the elevation model and set some values in a range to novalue:

```
block = QgsRasterBlock(dataType, cols, rows)

for row in range(rows):
    for col in range(cols):
        point = dtmLayer.dataProvider().transformCoordinates(QgsPoint(col, row), transformType)
        value, res = dtmLayer.dataProvider().sample(QgsPointXY(point.x(), point.y()), 1)
        
        finalValue = -9999.0
        if res and value != -9999.0:
            if value < 1500 or value > 1700:
                finalValue = value
        block.setValue(row, col, finalValue)
```

this sets every cell of the data block, resulting in the right map:

![image](https://github.com/qgis/QGIS/assets/390250/8bd79ccb-c772-4cb2-bbb9-af1454a5221a)

When doing a little change (indent that one line) to the above snippet, to set only certain values:
```
        finalValue = -9999.0
        if res and value != -9999.0:
            if value < 1500 or value > 1700:
                finalValue = value
                block.setValue(row, col, finalValue)
```

resulting in:

![image](https://github.com/qgis/QGIS/assets/390250/20da790f-6ffe-4ba5-8664-df110b5d13c3)

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.30.2-'s-Hertogenbosch | QGIS code revision | 0992b53397
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.6
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.37.2
PDAL version | 2.3.0
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Ubuntu 22.04.2 LTS
  |   |   |  
Active Python plugins
db_manager | 0.1.20
processing | 2.12.99
grassprovider | 2.12.99
MetaSearch | 0.3.6

</body></html>QGIS version
3.30.2-'s-Hertogenbosch
QGIS code revision
[0992b53397](https://github.com/qgis/QGIS/commit/0992b53397)
Qt version
5.15.3
Python version
3.10.6
GDAL/OGR version
3.4.1
PROJ version
8.2.1
EPSG Registry database version
v10.041 (2021-12-03)
GEOS version
3.10.2-CAPI-1.16.0
SQLite version
3.37.2
PDAL version
2.3.0
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.4
QScintilla2 version
2.11.6
OS version
Ubuntu 22.04.2 LTS

Active Python plugins
db_manager
0.1.20
processing
2.12.99
grassprovider
2.12.99
MetaSearch
0.3.6

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by nicogodet** [comment]
Could you please take a look to https://github.com/lutraconsulting/serval/tree/master/Serval ?
And check you are doing it right ?

**Comment by moovida** [comment]
@nicogodet , I guess you mean the raster_handler.py
In there it seems that it always works on all selected cell, therefore representing the usecase that also works for me.
Whereas setting just some of the values of the block results in distortions.

**Comment by agiudiceandrea** [comment]
@moovida, have you checked what is the content of a new QgsRasterBlock? Have you initialised it with some values?

**Comment by moovida** [comment]
@agiudiceandrea no, I did not and imagine that this is the issue. But I could not find any init method for that. So it all comes back to the need of setting all the cells of the block, right? That would solve the issue then.

Curiosity: is the pyQGIS API so C-near that when I create a QgsRasterBlock it contains whatever the allocated memory portion contains? No preinitialized something?

**Comment by agiudiceandrea** [comment]
> Curiosity: is the pyQGIS API so C-near that when I create a QgsRasterBlock it contains whatever the allocated memory portion contains? No preinitialized something?

It seems the case. I don't know if it is the correct behaviour or not. You could ask to the user or dev mailing list to check.
